org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTNode.java

```diff
@@ -2879,20 +2879,6 @@ final void checkModifiable() {
 
 	/**
      * Begin lazy initialization of this node.
-     * Here is the code pattern found in all AST
-     * node subclasses:
-     * <pre>
-     * if (this.foo == null) {
-	 *    // lazy init must be thread-safe for readers
-     *    synchronized (this) {
-     *       if (this.foo == null) {
-     *          preLazyInit();
-     *          this.foo = ...; // code to create new node
-     *          postLazyInit(this.foo, FOO_PROPERTY);
-     *       }
-     *    }
-     * }
-     * </pre>
```

Comment:
I didn't quite understand what's wrong with this code pattern, but if you change it, wouldn't it be nice to update this documentation, rather than removing?

Just recently I referred to this documentation in a review because I found it helpful.

Reply:
` this.foo = ...` writes to memory before executing postLazyInit, so any concurrent thread will see `this.foo != null` in the first check `if (this.foo == null) {` before entering synchronized. The other thread willl therefor not wait for postLazyInit to happen and may use a object that has is parent not initialized - if by chance postLazyInit() has not been executed yet.
I removed the documentation because it was a wrong pattern. All callers should now use same safe pattern where it can be copied from. But if you want i can also update the javadoc.

Reply:
> ` this.foo = ...` writes to memory before executing postLazyInit, so any concurrent thread will see `this.foo != null` in the first check `if (this.foo == null) {` before entering synchronized. The other thread willl therefor not wait for postLazyInit to happen and may use a object that has its parent not initialized - if by chance postLazyInit() has not been executed yet.

understood, thanks for explaining.

> I removed the documentation because it was a wrong pattern. All callers should now use same safe pattern where it can be copied from. But if you want i can also update the javadoc.

If we want all future additions to adhere to the new pattern, then documentation looks like a helpful device, no?

Reply:
good catch @jukzi ! 

Also, as @stephan-herrmann please update javadoc since DOM AST is always evolving and it would help in future commits. thanks!

Reply:
updated javadoc